# NumPy, Pandas, MLP

In [112]:
import numpy as np

### Задание 1
Создайте numpy array с элементами от числа N до 0 (например, для N = 10 это будет array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0]))

### Решение

In [113]:
N = 10

z = np.arange(0, N)
v1 = z[::-1]

print(v1, type(v1))

[9 8 7 6 5 4 3 2 1 0] <class 'numpy.ndarray'>


In [114]:
v2 = np.flip(z, 0)

print(v2)


[9 8 7 6 5 4 3 2 1 0]


### Задание 2
Создайте диагональную матрицу с элементами от N до 0. Посчитайте сумму ее значений на диагонали.

### Решение

In [116]:
N = 10

y = np.diag(np.flip(np.arange(0, N), 0), k=0)
print(y)

[[9 0 0 0 0 0 0 0 0 0]
 [0 8 0 0 0 0 0 0 0 0]
 [0 0 7 0 0 0 0 0 0 0]
 [0 0 0 6 0 0 0 0 0 0]
 [0 0 0 0 5 0 0 0 0 0]
 [0 0 0 0 0 4 0 0 0 0]
 [0 0 0 0 0 0 3 0 0 0]
 [0 0 0 0 0 0 0 2 0 0]
 [0 0 0 0 0 0 0 0 1 0]
 [0 0 0 0 0 0 0 0 0 0]]


In [117]:
y.sum()

45

In [118]:
np.trace(y)

45

### Задание 3
Решите систему уравнений:

4x + 2y + z = 4

x + 3y = 12

5y + 4z = -3

### Решение

In [119]:
a = np.array([[4, 2, 1], [1, 3, 0], [0, 5, 4]])
b = np.array([4, 12, -3])

In [120]:
from numpy import linalg

linalg.solve(a, b)

array([ 0.46666667,  3.84444444, -5.55555556])

In [121]:
np.allclose(np.dot(a, linalg.solve(a, b)), b)

True

### Задание 4

Имеется матрица покупок в интернет-магазине. Столбец А - ID пользователя. Остальные столбцы - количество покупок категорий товаров этим пользователем:

Матрица в виде numpy array

In [123]:
users_stats = np.array(
    [
        [2, 1, 0, 0, 0, 0],
        [1, 1, 2, 1, 0, 0],
        [2, 0, 1, 0, 0, 0],
        [1, 1, 2, 1, 0, 1],
        [0, 0, 1, 2, 0, 0],
        [0, 0, 0, 0, 0, 5],
        [1, 0, 0, 0, 0, 0],
        [0, 1, 1, 0, 0, 0],
        [0, 0, 0, 1, 1, 3],
        [1, 0, 0, 2, 1, 4]
    ], 
    np.int32
)

На сайт заходит очередной посетитель, о покупках которого известно следующее:

In [124]:
next_user_stats = np.array([0, 1, 2, 0, 0, 0])

Найдите самого похожего пользователя. Т. е. посчитайте косинусное сходство между этим пользователем и всеми пользователями из массива user_stats

### Решение

In [126]:
res = {}
for user_stats in users_stats:
    """
    Подсчет косинуса угла между векторами a, b по их координатам
    """
    # длины векторов
    aLength = np.linalg.norm( user_stats )
    bLength = np.linalg.norm( next_user_stats )
    cosine = round(np.dot( user_stats, next_user_stats ) / ( aLength * bLength ), 2)
    res[str(user_stats), str(next_user_stats)] = float(cosine) # создание результирующщего словаря из которого будет найдено max значение косинусного сходства между user'ами
res

{('[2 1 0 0 0 0]', '[0 1 2 0 0 0]'): 0.2,
 ('[1 1 2 1 0 0]', '[0 1 2 0 0 0]'): 0.85,
 ('[2 0 1 0 0 0]', '[0 1 2 0 0 0]'): 0.4,
 ('[1 1 2 1 0 1]', '[0 1 2 0 0 0]'): 0.79,
 ('[0 0 1 2 0 0]', '[0 1 2 0 0 0]'): 0.4,
 ('[0 0 0 0 0 5]', '[0 1 2 0 0 0]'): 0.0,
 ('[1 0 0 0 0 0]', '[0 1 2 0 0 0]'): 0.0,
 ('[0 1 1 0 0 0]', '[0 1 2 0 0 0]'): 0.95,
 ('[0 0 0 1 1 3]', '[0 1 2 0 0 0]'): 0.0,
 ('[1 0 0 2 1 4]', '[0 1 2 0 0 0]'): 0.0}

In [127]:
# вывод пользователей между векторами которых косинусное сходство максимально приближено к 1

max_val = max(res.values())
final_dict = {k:v for k, v in res.items() if v == max_val}
final_dict

{('[0 1 1 0 0 0]', '[0 1 2 0 0 0]'): 0.95}

In [128]:
import pandas as pd

### Задание 5

Скачайте с сайта https://grouplens.org/datasets/movielens/ датасет любого размера. Определите какому фильму было выставлено больше всего оценок 5.0.

### Решение

In [132]:
ratings = pd.read_csv('C:/Netology/ml-25m/ratings.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1      296     5.0  1147880044
1       1      306     3.5  1147868817
2       1      307     5.0  1147868828
3       1      665     5.0  1147878820
4       1      899     3.5  1147868510

In [133]:
"""В таблице top_movies только фильмы у которых оценки пользователей равны 5.0,
используем метод .value_counts(), с помощью этого метода мы узнаем какое количество раз 
встречается movieId в датафрейме top_movies,
так как метод .value_counts() использует сортировку по убыванию, 
то первый в списке фильм и будет ответом задание."""

movies = ratings[['movieId', 'rating']] # таблица с фильмами и их рейтингом
top_movies = movies.query(f'rating == 5.0') # таблица с фильмами, оценки которых равны 5.0
top_movies['movieId'].value_counts().head(10) # вывод ТОП-10 фильмов которым выставлены оценки 5.0

318     39553
296     32169
356     25918
260     25804
2571    25482
527     24853
593     24801
858     24418
50      21585
2959    21486
Name: movieId, dtype: int64

## Ответ 
Фильм с ID равному 318 выставлено больше всего оценок 5.0

In [137]:
tab_movies = pd.read_csv('C:/Netology/ml-25m/movies.csv')

In [138]:
joined = ratings.merge(tab_movies, on='movieId', how='left')
joined.head()

userId  movieId  rating   timestamp  \
0       1      296     5.0  1147880044   
1       1      306     3.5  1147868817   
2       1      307     5.0  1147868828   
3       1      665     5.0  1147878820   
4       1      899     3.5  1147868510   

                                              title  \
0                               Pulp Fiction (1994)   
1  Three Colors: Red (Trois couleurs: Rouge) (1994)   
2  Three Colors: Blue (Trois couleurs: Bleu) (1993)   
3                                Underground (1995)   
4                        Singin' in the Rain (1952)   

                        genres  
0  Comedy|Crime|Drama|Thriller  
1                        Drama  
2                        Drama  
3             Comedy|Drama|War  
4       Comedy|Musical|Romance

In [140]:
movies = joined[['movieId', 'title', 'rating']] # таблица с фильмами и их рейтингом
top_movies = movies.query(f'rating == 5.0') # таблица с фильмами, оценки которых равны 5.0
top_movies['title'].value_counts().head(10) # вывод ТОП-10 фильмов которым выставлены оценки 5.0

Shawshank Redemption, The (1994)             39553
Pulp Fiction (1994)                          32169
Forrest Gump (1994)                          25918
Star Wars: Episode IV - A New Hope (1977)    25804
Matrix, The (1999)                           25482
Schindler's List (1993)                      24853
Silence of the Lambs, The (1991)             24801
Godfather, The (1972)                        24418
Usual Suspects, The (1995)                   21585
Fight Club (1999)                            21486
Name: title, dtype: int64

## Ответ

Фильму Shawshank Redemption, The (1994) выставлено больше всего оценок 5.0

## Проверка

In [141]:
joined.query(f'movieId == 318').head(1)

userId  movieId  rating   timestamp                             title  \
79       2      318     5.0  1141417181  Shawshank Redemption, The (1994)   

         genres  
79  Crime|Drama

### Задание 6
По данным файла power.csv посчитайте суммарное потребление стран Прибалтики (Латвия, Литва и Эстония) категорий 4, 12 и 21 за период с 2005 по 2010 года. Не учитывайте в расчетах отрицательные значения quantity.

### Решение 

In [143]:
countries = pd.read_csv('C:/Netology/Lecture_notebook/10._Basics_Pandas/power.csv')
countries.head()

country  year  quantity  category
0  Austria  1996       5.0         1
1  Austria  1995      17.0         1
2  Belgium  2014       0.0         1
3  Belgium  2013       0.0         1
4  Belgium  2012      35.0         1

In [144]:
# создание датафрейма с тремя странами Прибалтики (Латвия, Литва, Эстония)

baltics_countries = countries[(countries['country']=='Latvia') | (countries['country']=='Lithuania') | (countries['country']=='Estonia')]
baltics_countries.head()

country  year  quantity  category
373  Lithuania  2014      39.0         1
374  Lithuania  2013     192.0         1
375  Lithuania  2012     196.0         1
376  Lithuania  2011     171.0         1
377  Lithuania  2010     158.0         1

In [145]:
# фильтр датафрейма стран Прибалтики с учетом категорий, вывод только категорий 4, 12 и 21
# фильтр датафрейма стран Прибалтики с учетом положительного значение quantity (quantity = 0 или quantity > 0)
# фильтр датафрейма стран Прибалтики с года (с 2005 по 2010 включительно)


filtered_baltics_countries = baltics_countries[((baltics_countries['category'] == 4) | (baltics_countries['category'] == 12) | (baltics_countries['category'] == 21)) & ((baltics_countries['quantity'] > 0) | (baltics_countries['quantity'] == 0))].query(f'2005 <= year <= 2010')
filtered_baltics_countries.sort_values(by=['year', 'country', 'quantity'], ascending=[True, True, False]).head(20)

country  year  quantity  category
247355  Estonia  2005    2559.0        21
241834  Estonia  2005    2527.0        21
232015  Estonia  2005    2523.0        21
209277  Estonia  2005    2492.0        21
227717  Estonia  2005      32.0        21
221417  Estonia  2005      31.0        21
253566  Estonia  2005      31.0        21
219383  Estonia  2005      30.0        21
19857   Estonia  2005       6.0         4
21903   Estonia  2005       5.0         4
214747  Estonia  2005       5.0        21
237507  Estonia  2005       5.0        21
10620   Estonia  2005       1.0         4
12758   Estonia  2005       1.0         4
15528   Estonia  2005       1.0         4
17678   Estonia  2005       1.0         4
25746   Estonia  2005       1.0         4
226317  Estonia  2005       1.0        21
224567  Estonia  2005       0.0        21
248555   Latvia  2005    2166.0        21

In [146]:
# суммирование столбца quantity
# группировка по столбцам country, year, category
# сортировка по year, country, category

filtered_baltics_countries.groupby(by=['country','year', 'category']).sum().sort_values(by=['year', 'country', 'category'], ascending=[True, True, True]).head(60)

quantity
country   year category          
Estonia   2005 4             16.0
               21         10236.0
Latvia    2005 21          8664.0
Lithuania 2005 4              6.0
               12            18.0
               21         18224.0
Estonia   2006 4              6.0
               21         10228.0
Latvia    2006 21          8604.0
Lithuania 2006 4              6.0
               12            18.0
               21         19868.0
Estonia   2007 4              0.0
               21         10508.0
Latvia    2007 4              0.0
               21          8532.0
Lithuania 2007 4              0.0
               12          2454.0
               21         19972.0
Estonia   2008 4              6.0
               21         10776.0
Latvia    2008 4              0.0
               21          8620.0
Lithuania 2008 4              6.0
               12          2029.0
               21         20140.0
Estonia   2009 21         10660.0
Latvia    2009 4              0.0
               21         10104.0
Lithuania 2009 12           862.0
               21         21156.0
Estonia   2010 21         11004.0
Latvia    2010 4              0.0
               21         10320.0
Lithuania 2010 12           945.0
               21         16592.0

### Задание 7
Выберите страницу любого сайта с табличными данными. Импортируйте таблицы в pandas dataframe.
Примеры страниц (необязательно брать именно эти):

### Решение

In [147]:
data = pd.read_html('https://www.bbc.com/russian/news-51706538')
data[0][0:20]

Unnamed: 0   Умерли  Смертность*  Заразились  \
0              США  290 803        889.0  15 548 135   
1         Бразилия  179 765        858.0   6 781 799   
2            Индия  142 186        105.0   9 796 769   
3          Мексика  112 326        890.0   1 217 126   
4   Великобритания   63 082        940.0   1 787 783   
5           Италия   62 626       1033.0   1 787 147   
6          Франция   56 940        876.0   2 337 966   
7             Иран   51 496        630.0   1 083 023   
8          Испания   47 344       1014.0   1 720 056   
9           Россия   45 370        311.0   2 574 319   
10       Аргентина   40 431        911.0   1 482 216   
11        Колумбия   38 484        775.0   1 399 911   
12            Перу   36 499       1141.0     979 111   
13             ЮАР   22 747        394.0     836 764   
14          Польша   21 630        570.0   1 102 096   
15        Германия   21 064        253.0   1 287 092   
16       Индонезия   18 336         69.0     598 933   
17         Бельгия   17 692       1541.0     600 397   
18            Чили   15 774        842.0     566 440   
19          Турция   15 751        191.0     955 766   

    Новые случаи  0  10  100  1000  10000  **  Unnamed: 5  
0                                         NaN         NaN  
1                                         NaN         NaN  
2                                         NaN         NaN  
3                                         NaN         NaN  
4                                         NaN         NaN  
5                                         NaN         NaN  
6                                         NaN         NaN  
7                                         NaN         NaN  
8                                         NaN         NaN  
9                                         NaN         NaN  
10                                        NaN         NaN  
11                                        NaN         NaN  
12                                        NaN         NaN  
13                                        NaN         NaN  
14                                        NaN         NaN  
15                                        NaN         NaN  
16                                        NaN         NaN  
17                                        NaN         NaN  
18                                        NaN         NaN  
19                                        NaN         NaN

In [148]:
data[0][0:2]

Unnamed: 0   Умерли  Смертность*  Заразились  \
0        США  290 803        889.0  15 548 135   
1   Бразилия  179 765        858.0   6 781 799   

   Новые случаи  0  10  100  1000  10000  **  Unnamed: 5  
0                                        NaN         NaN  
1                                        NaN         NaN

In [149]:
data[0][0:1]

Unnamed: 0   Умерли  Смертность*  Заразились  \
0        США  290 803        889.0  15 548 135   

   Новые случаи  0  10  100  1000  10000  **  Unnamed: 5  
0                                        NaN         NaN

### Задание 8

Задание 1**
Напишите функцию, которая классифицирует фильмы из материалов занятия по следующим правилам:
- оценка 2 и меньше - низкий рейтинг
- оценка 4 и меньше - средний рейтинг
- оценка 4.5 и 5 - высокий рейтинг

Результат классификации запишите в столбец class

### Решение

In [150]:
# открываем и читаем таблицу с рейтингами

ratings = pd.read_csv('C:/Netology/ml-25m/ratings.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1      296     5.0  1147880044
1       1      306     3.5  1147868817
2       1      307     5.0  1147868828
3       1      665     5.0  1147878820
4       1      899     3.5  1147868510

In [151]:
# считаем средний рейтинг для каждого фильма и формируем таблицу со средним рейтингом

average_rating = ratings[['movieId','rating']].groupby('movieId').mean().round(2)
average_rating.head()

rating
movieId        
1          3.89
2          3.25
3          3.14
4          2.85
5          3.06

In [152]:
# функция присваивающая классификацию фильма на основе его рейтинг

def class_movie(rating):
    if rating <= 2:
        return 'низкий рейтинг'
    elif 2 < rating <= 4:
        return  'средний рейтинг'
    elif rating == 4.5 or rating == 5:
        return 'высокий рейтинг'

In [153]:
# применяем функцию к датафрейму со средним рейтигом фильмов

average_rating['class'] = average_rating['rating'].apply(class_movie)
average_rating.head()

rating            class
movieId                         
1          3.89  средний рейтинг
2          3.25  средний рейтинг
3          3.14  средний рейтинг
4          2.85  средний рейтинг
5          3.06  средний рейтинг

In [154]:
average_rating.sample(5)

rating            class
movieId                         
195299     4.00  средний рейтинг
42721      1.68   низкий рейтинг
3287       2.90  средний рейтинг
149086     3.00  средний рейтинг
154578     1.67   низкий рейтинг

### Задание 9

Используем файл keywords.csv.

Необходимо написать гео-классификатор, который каждой строке сможет выставить географическую принадлежность определенному региону. Т. е. если поисковый запрос содержит название города региона, то в столбце ‘region’ пишется название этого региона. Если поисковый запрос не содержит названия города, то ставим ‘undefined’.

Правила распределения по регионам Центр, Северо-Запад и Дальний Восток:

In [155]:
geo_data = {
'Центр': ['москва', 'тула', 'ярославль'],
'Северо-Запад': ['петербург', 'псков', 'мурманск'],
'Дальний Восток': ['владивосток', 'сахалин', 'хабаровск']
}

### Решение

In [157]:
keywords = pd.read_csv('C:/Netology/keywords.csv')
keywords.head()

keyword     shows
0             вк  64292779
1  одноклассники  63810309
2          порно  41747114
3           ютуб  39995567
4      вконтакте  21014195

In [158]:
# функция которая ищет название города в запросе пользователя

def geo_classifier(keyword):
    for key, value in geo_data.items(): #используя цикл для того чтобы добраться до названия города и присвоеть его переменной City
        for city in value:
            if city in keyword.lower().split(' '):
                return key
            
    return 'undefined'

In [159]:
# применяем функцию к датафрейму с поисковыми запросами

keywords['region'] = keywords['keyword'].apply(geo_classifier)
keywords.sample(10)

keyword  shows     region
53122               мосуслуги pgu mos ru личный кабинет   6577  undefined
70167                                    новости пскова   5036  undefined
59452                                   даниил вахрушев   5934  undefined
39894                                            малуша   8412  undefined
59220                             дружба народов сериал   6217  undefined
74367                                        айфон 5 se   4836  undefined
32744  дом странных детей мисс перегрин скачать торрент  11008  undefined
33825                                              джек   9743  undefined
76971                                       контрибуция   4634  undefined
41878                                          inkscape   8055  undefined

### Проверка количества поисковых запросов содержащих города регионов

In [160]:
# делаем проверку чтобы найти поисковые запросы с указанием городов регионов

keywords.groupby('region').count().reset_index()[['region', 'keyword']]

region  keyword
0       undefined    99406
1  Дальний Восток       64
2    Северо-Запад      146
3           Центр      384

### Задание 10

Есть мнение, что “раньше снимали настоящее кино, не то что сейчас”. Ваша задача проверить это утверждение, используя файлы с рейтингами фильмов из прошлого домашнего занятия (файл ratings.csv из базы https://grouplens.org/datasets/movielens). Т. е. проверить верно ли, что с ростом года выпуска фильма его средний рейтинг становится ниже.

При этом мы не будем затрагивать субьективные факторы выставления этих рейтингов, а пройдемся по следующему алгоритму:

В переменную years запишите список из всех годов с 1950 по 2010.

Напишите функцию production_year, которая каждой строке из названия фильма выставляет год выпуска. Не все названия фильмов содержат год выпуска в одинаковом формате, поэтому используйте следующий алгоритм:

для каждой строки пройдите по всем годам списка years
если номер года присутствует в названии фильма, то функция возвращает этот год как год выпуска
если ни один из номеров года списка years не встретился в названии фильма, то возвращается 1900 год
Запишите год выпуска фильма по алгоритму пункта 2 в новый столбец ‘year’

Посчитайте средний рейтинг всех фильмов для каждого значения столбца ‘year’ и отсортируйте результат по убыванию рейтинга

In [161]:
# датафрейм с фильмами

tab_movies = pd.read_csv('C:/Netology/ml-25m/movies.csv')
tab_movies.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [162]:
# список годов от 1950 по 2010

years = list(range(1950, 2011))

In [163]:
# функция ищет год в названии фильма

def production_year(title):
    for year in years:  #через цикл проверяем каждый год на наличие в названии фильма
        if str(year) in title:
            #print(title)
            #print(year)
            return str(year)
    return '1900'

In [164]:
# применяем функцию к датафрейму

tab_movies['year'] = tab_movies['title'].apply(production_year)
tab_movies.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  year  
0  Adventure|Animation|Children|Comedy|Fantasy  1995  
1                   Adventure|Children|Fantasy  1995  
2                               Comedy|Romance  1995  
3                         Comedy|Drama|Romance  1995  
4                                       Comedy  1995

In [166]:
# фильтруем нужные нам столбцы с учёт периода с 1950 по 2010 года

movies_year = tab_movies[['movieId', 'title', 'year']][(tab_movies['year'] >= '1950') & (tab_movies['year'] <= '2010')]
movies_year.head()

movieId                               title  year
0        1                    Toy Story (1995)  1995
1        2                      Jumanji (1995)  1995
2        3             Grumpier Old Men (1995)  1995
3        4            Waiting to Exhale (1995)  1995
4        5  Father of the Bride Part II (1995)  1995

In [168]:
# датафрейм со средним рейтингом фильмов

average_rating = ratings[['movieId','rating']].groupby('movieId').mean().round(2)
average_rating.head()

rating
movieId        
1          3.89
2          3.25
3          3.14
4          2.85
5          3.06

In [169]:
# соединяем два датафрейма в один отфильтрованные по годам фильмы и их средний рейтинг, сортируем по убыванию рейтинга

joined = movies_year.merge(average_rating, on='movieId', how='left')
joined.sort_values('rating', ascending=False).head(15)

movieId                                              title  year  \
29687   164620                               Zig Zag Story (1983)  1983   
19839   124482                          The Heartbreak Kid (1993)  1993   
32576   178217                   Hempsters: Plant the Seed (2008)  2008   
19640   123771                           The Cariboo Trail (1950)  1950   
32608   178417                               A Lustful Man (1961)  1961   
8726     27914  Hijacking Catastrophe: 9/11, Fear & the Sellin...  2004   
36396   199762                        Der letzte Fußgänger (1960)  1960   
19701   123959                                     Groupie (2010)  2010   
36393   199756                    Zum Teufel mit der Penne (1968)  1968   
32671   178739                                Kiss Me Kate (2003)  2003   
19751   124174                             Until They Sail (1957)  1957   
24581   143422                                           2 (2007)  2007   
24545   143269                           Moro Witch Doctor (1964)  1964   
24539   143235                            Going to the Mat (2004)  2004   
24535   143223      It Seemed Like a Good Idea at the Time (1975)  1975   

       rating  
29687     5.0  
19839     5.0  
32576     5.0  
19640     5.0  
32608     5.0  
8726      5.0  
36396     5.0  
19701     5.0  
36393     5.0  
32671     5.0  
19751     5.0  
24581     5.0  
24545     5.0  
24539     5.0  
24535     5.0

## Проверка нашей гипотезы:
Опираясь на вывод первых 15 строк,в нашем веке также есть достойные и стоящие просмотра фильмы. 

In [170]:
from pymystem3 import Mystem

### Задание 11

Для датафрейма log из материалов занятия создайте столбец source_type по следующим правилам:

- если источник traffic_source равен yandex или google, то в source_type ставится organic
- для источников paid и email из России - ставим ad
- для источников paid и email не из России - ставим other
- все остальные варианты берем из traffic_source без изменений

### Решение

In [173]:
log = pd.read_csv('C:/Netology/Python_12_pivot_and_str/visit_log.csv', sep=';')
log.head()

timestamp    visit_id                                 url  region  \
0  1549980692  e3b0c44298  https://host.ru/3c19b4ef7371864fa3  Russia   
1  1549980704  6e340b9cff  https://host.ru/c8d9213a31839f9a3a  Russia   
2  1549980715  96a296d224  https://host.ru/b8b58337d272ee7b15  Russia   
3  1549980725  709e80c884  https://host.ru/b8b58337d272ee7b15  Russia   
4  1549980736  df3f619804  https://host.ru/b8b58337d272ee7b15  Russia   

      user_id traffic_source  
0  b1613cc09f         yandex  
1  4c3ec14bee         direct  
2  a8c40697fb         yandex  
3  521ac1d6a0         yandex  
4  d7323c571c         yandex

In [174]:
log.loc[log.traffic_source.isin(['yandex', 'google']), 'source_type'] = 'organic'
log.loc[log.traffic_source.isin(['paid', 'email']) & log.region.isin(['Russia']), 'source_type'] = 'ad'
log.loc[(log.traffic_source.isin(['paid', 'email'])) & (log.region != 'Russia'), 'source_type'] = 'other'
log.loc[(~log.traffic_source.isin(['yandex', 'google', 'paid', 'email'])) & ((log.region != 'Russia') | (log.region.isin(['Russia']))), 'source_type'] = log['traffic_source']
log.head(20)

timestamp    visit_id                                 url   region  \
0   1549980692  e3b0c44298  https://host.ru/3c19b4ef7371864fa3   Russia   
1   1549980704  6e340b9cff  https://host.ru/c8d9213a31839f9a3a   Russia   
2   1549980715  96a296d224  https://host.ru/b8b58337d272ee7b15   Russia   
3   1549980725  709e80c884  https://host.ru/b8b58337d272ee7b15   Russia   
4   1549980736  df3f619804  https://host.ru/b8b58337d272ee7b15   Russia   
5   1549980742  8855508aad  https://host.ru/df646c3676cc259fa0   Russia   
6   1549980742  b0f66adc83  https://host.ru/b8b58337d272ee7b15   Russia   
7   1549980754  837885c8f8  https://host.ru/108ce4b365afb7b88e   Russia   
8   1549980760  af5570f5a1  https://host.ru/3004a8273caeef2867    China   
9   1549980765  3e7077fd2f  https://host.ru/df646c3676cc259fa0   Russia   
10  1549980777  01d448afd9  https://host.ru/a5dda93e70318570c0    India   
11  1549980785  71b6c1d538  https://host.ru/c8d9213a31839f9a3a  Ukraine   
12  1549980789  15ec7bf0b5  https://host.ru/8e88d4703848cc0ec4   Russia   
13  1549980792  dd46c3eebb  https://host.ru/3004a8273caeef2867    China   
14  1549980793  e7ecebbc59  https://host.ru/05e41944dd85a900a6  Belarus   
15  1549980797  5322fecfc9  https://host.ru/4adc8a29b212051b04  Belarus   
16  1549980801  374708fff7  https://host.ru/39fa884393666d45fc   Russia   
17  1549980807  0a88111852  https://host.ru/2fcb223266880ccf57  Ukraine   
18  1549980814  60daa3a5f7  https://host.ru/ef57543002ce2a0f2a  Belarus   
19  1549980816  d6fd62f5ce  https://host.ru/f1eb4601740d627ab0  Belarus   

       user_id traffic_source source_type  
0   b1613cc09f         yandex     organic  
1   4c3ec14bee         direct      direct  
2   a8c40697fb         yandex     organic  
3   521ac1d6a0         yandex     organic  
4   d7323c571c         yandex     organic  
5   fc43898e47         yandex     organic  
6   13fc55e781           paid          ad  
7   cb5082b6f6         direct      direct  
8   45664f7af2         direct      direct  
9   6f9de8c8b6          email          ad  
10  27e76a4a80         direct      direct  
11  3b13701488         google     organic  
12  6d1d81f7f8           paid          ad  
13  7e301c8c41         direct      direct  
14  ac52dedc3b         direct      direct  
15  9a45d98867           paid       other  
16  e9bd59e22d         google     organic  
17  a2ba8fa3db           paid       other  
18  72cb10fb5c         direct      direct  
19  7269dcd869         direct      direct

### Задание 12
В файле URLs.txt содержатся url страниц новостного сайта. Вам необходимо отфильтровать его по адресам страниц с текстами новостей. Известно, что шаблон страницы новостей имеет внутри url следующую конструкцию: /, затем 8 цифр, затем дефис. Выполните следующие действия:

Прочитайте содержимое файла с датафрейм
Отфильтруйте страницы с текстом новостей, используя метод str.contains и регулярное выражение в соответствии с заданным шаблоном

### Решение

In [178]:
news = pd.read_csv('C:/Netology/Python_12_pivot_and_str/URLs.txt')
news.head()

url
0                                            /world/
1                                           /latest/
2                                      /?updated=top
3  /politics/36188461-s-marta-zhizn-rossiyan-susc...
4  /world/36007585-tramp-pridumal-kak-reshit-ukra...

In [179]:
import re

news[news.url.str.contains(r'/[0-9]{8}-')].head()

url
3  /politics/36188461-s-marta-zhizn-rossiyan-susc...
4  /world/36007585-tramp-pridumal-kak-reshit-ukra...
5  /science/36157853-nasa-sobiraet-ekstrennuyu-pr...
6  /video/36001498-poyavilis-pervye-podrobnosti-g...
7  /world/36007585-tramp-pridumal-kak-reshit-ukra...

### Задание 13
В датафрейме data создайте столбец lemmas, в котором вычислите леммы поисковых запросов из столбца keyword. Леммы должны иметь строковый тип.

### Решение

In [180]:
data = pd.DataFrame({
    'keyword': ['курс гривны к рублю', 'доллары в рубли', '100 долларов в рублях', 'курс рубля'],
    'shows': [125076, 114173, 97534, 53546],
})

In [181]:
def lemmas(keyword):
    m = Mystem()
    lemmas = m.lemmatize(keyword)
    return ''.join(lemmas).strip()
    

In [182]:
data['lemmas'] = data['keyword'].apply(lemmas)
data

keyword   shows               lemmas
0    курс гривны к рублю  125076  курс гривна к рубль
1        доллары в рубли  114173       доллар в рубль
2  100 долларов в рублях   97534   100 доллар в рубль
3             курс рубля   53546           курс рубль

### Задание 14
Используйте файл с оценками фильмов ml-latest-small/ratings.csv. Посчитайте среднее время жизни пользователей, которые выставили более 100 оценок. Под временем жизни понимается разница между максимальным и минимальным значением столбца timestamp для данного значения userId.

### Решение

In [183]:
ratings = pd.read_csv('C:/Netology/ml-latest-small/ratings.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [184]:
my_table = ratings.filter(items = ['userId'])
my_table.head()

userId
0       1
1       1
2       1
3       1
4       1

In [185]:
my_table.drop_duplicates(subset = 'userId', keep = 'first', inplace = True)
my_table.head()

userId
0         1
20        2
96        3
147       4
351       5

In [186]:
max_time = ratings.groupby('userId').agg(['min', 'max'])['timestamp']['max']
max_time.head(10)

userId
1     1260759205
2      835356246
3     1298932787
4      949982274
5     1163375145
6     1109258285
7      851869291
8     1154474527
9      939122916
10     942767571
Name: max, dtype: int64

In [187]:
min_time = ratings.groupby('userId').agg(['min', 'max'])['timestamp']['min']
min_time.head(10)

userId
1     1260759108
2      835355395
3     1298861589
4      949778714
5     1163373044
6     1108134263
7      851866703
8     1154389340
9      938627748
10     942765978
Name: min, dtype: int64

In [188]:
joined_max = pd.merge(my_table, max_time, on='userId', how='left')
joined_max.head()

userId         max
0       1  1260759205
1       2   835356246
2       3  1298932787
3       4   949982274
4       5  1163375145

In [189]:
joined_max_min = pd.merge(joined_max, min_time, on='userId', how='left')
joined_max_min.head()

userId         max         min
0       1  1260759205  1260759108
1       2   835356246   835355395
2       3  1298932787  1298861589
3       4   949982274   949778714
4       5  1163375145  1163373044

In [190]:
joined_max_min['lifetime'] = joined_max_min['max'] - joined_max_min['min']
joined_max_min.head()

userId         max         min  lifetime
0       1  1260759205  1260759108        97
1       2   835356246   835355395       851
2       3  1298932787  1298861589     71198
3       4   949982274   949778714    203560
4       5  1163375145  1163373044      2101

In [191]:
users_grouped = ratings.groupby('userId').count()
users_grouped.rename(columns={'movieId': 'quantity_rating'}, inplace=True)
quantity_rating = users_grouped.filter(items = ['quantity_rating'])
quantity_rating.reset_index().head()

userId  quantity_rating
0       1               20
1       2               76
2       3               51
3       4              204
4       5              100

In [193]:
lifetime_table = pd.merge(joined_max_min, quantity_rating, on='userId', how='left')
result_table =  lifetime_table.loc[lifetime_table.quantity_rating > 100]
result_table.head()

userId         max         min   lifetime  quantity_rating
3        4   949982274   949778714     203560              204
7        8  1154474527  1154389340      85187              116
14      15  1469330735   997937239  471393496             1700
16      17  1127476640  1127468587       8053              363
18      19   855195373   855190091       5282              423

In [194]:
sum(result_table['lifetime'])

10340770922

In [195]:
result_table['userId'].count()

258

In [196]:
day = 86400 # 1 день равен 86400 секунд

mean_lifetime = round((sum(result_table['lifetime']) / result_table['userId'].count()) / day, 0)
mean_lifetime

464.0

### Ответ:
Среднее время жизни пользователей, которые выставили более 100 оценок равно 464 дня

### Задание 15

Дана статистика услуг перевозок клиентов компании по типам (см. файл с кодом занятия). Необходимо сформировать две таблицы:
таблицу с тремя типами выручки для каждого client_id без указания адреса клиента
аналогичную таблицу по типам выручки с указанием адреса клиента
Обратите внимание, что в процессе объединения таблиц данные не должны теряться.

Дана статистика услуг перевозок клиентов компании по типам:

- rzd - железнодорожные перевозки
- auto - автомобильные перевозки
- air - воздушные перевозки
- client_base - адреса клиентов

### Решение

In [197]:
rzd = pd.DataFrame(
    {
        'client_id': [111, 112, 113, 114, 115],
        'rzd_revenue': [1093, 2810, 10283, 5774, 981]
    }
)
rzd

client_id  rzd_revenue
0        111         1093
1        112         2810
2        113        10283
3        114         5774
4        115          981

In [198]:
auto = pd.DataFrame(
    {
        'client_id': [113, 114, 115, 116, 117],
        'auto_revenue': [57483, 83, 912, 4834, 98]
    }
)
auto

client_id  auto_revenue
0        113         57483
1        114            83
2        115           912
3        116          4834
4        117            98

In [199]:
air = pd.DataFrame(
    {
        'client_id': [115, 116, 117, 118],
        'air_revenue': [81, 4, 13, 173]
    }
)
air

client_id  air_revenue
0        115           81
1        116            4
2        117           13
3        118          173

In [200]:
client_base = pd.DataFrame(
    {
        'client_id': [111, 112, 113, 114, 115, 116, 117, 118],
        'address': ['Комсомольская 4', 'Энтузиастов 8а', 'Левобережная 1а', 'Мира 14', 'ЗЖБИиДК 1', 
                    'Строителей 18', 'Панфиловская 33', 'Мастеркова 4']
    }
)
client_base

client_id          address
0        111  Комсомольская 4
1        112   Энтузиастов 8а
2        113  Левобережная 1а
3        114          Мира 14
4        115        ЗЖБИиДК 1
5        116    Строителей 18
6        117  Панфиловская 33
7        118     Мастеркова 4

In [201]:
info_client = client_base.filter( items = ['client_id'])
info_client.head(10)

client_id
0        111
1        112
2        113
3        114
4        115
5        116
6        117
7        118

In [202]:
joined_client_1 = (
    info_client
    .merge(rzd, on='client_id', how='left')
    .merge(auto, on='client_id', how='left')
    .merge(air, on='client_id', how='left')
)
joined_client_1.fillna(0).head(10)

client_id  rzd_revenue  auto_revenue  air_revenue
0        111       1093.0           0.0          0.0
1        112       2810.0           0.0          0.0
2        113      10283.0       57483.0          0.0
3        114       5774.0          83.0          0.0
4        115        981.0         912.0         81.0
5        116          0.0        4834.0          4.0
6        117          0.0          98.0         13.0
7        118          0.0           0.0        173.0

In [203]:
joined_client_2 = (
    client_base
    .merge(rzd, on='client_id', how='left')
    .merge(auto, on='client_id', how='left')
    .merge(air, on='client_id', how='left')
)
joined_client_2.fillna(0).head(10)

client_id          address  rzd_revenue  auto_revenue  air_revenue
0        111  Комсомольская 4       1093.0           0.0          0.0
1        112   Энтузиастов 8а       2810.0           0.0          0.0
2        113  Левобережная 1а      10283.0       57483.0          0.0
3        114          Мира 14       5774.0          83.0          0.0
4        115        ЗЖБИиДК 1        981.0         912.0         81.0
5        116    Строителей 18          0.0        4834.0          4.0
6        117  Панфиловская 33          0.0          98.0         13.0
7        118     Мастеркова 4          0.0           0.0        173.0

### Задание 16
В задаче сквозной аналитики вам предоставили данные по местоположению пользователей. Т. е. для каждого user_id известна последовательность координат (широта/долгота), когда они требовались приложению для полноценной работы. Как бы вы добавили эти сведения в таблицу визитов и покупок? Для составления ответа можно использовать вопросы:

- У каждого пользователя известен набор координат. А для связывания с визитом или фактом покупки скорее всего потребуется одно-два числа. Как их получить?
- Наборы координат одного и того же пользователя могут быть значительно удалены друг от друга. Как это отразится на вопросе расчетах пункта 1?
- Какие дополнительные признаки можно получить из координат? Ведь это просто числа, которые сами по себе мало что дают.

### Размышления:

Координаты являются ункальными значениями, один адрес пользователя, одно его местоположение это две координаты, у пользователя может быть один или несколько адресов, то есть при добавлении координат в таблица визитов и покупок, по итогу мы должны получить строку: user_id, визиты, покупки, адрес(или координаты), где user_id уникален, но может повторяться в строках (в зависимости от количества адресов), визиты и покупки мы можем суммировать, возможно мы можем разместить в строке один адрес, а можем список адресов одного пользователя, тогда строка будет выглядеть так: user_id, sum(визитов), sum(покупок), список адресов, то есть допускаем и тот и тот вариант, в зависимости от наших целей.

Если у пользователя несколько местоположений, то в таком случаем из координат нескольких местоположений пользователя, можно составить полигон, вершинами которого будут координаты местоположений пользователя, тогда наша строка будет как: user_id, sum(визитов), sum(покупок), полигон.

Связывание набора координат и визиты и покупки пользователя, будет через user_id, который у каждого пользователя уникален.
Иначе не совсем понятен вопрос, о какой взаимосвязи речь, о той которая поможет объединить данные(таблицы), или как связаны визиты и покупки с координатами (думается, что никак)?

При расчете пункта один, если у пользователя несколько местоположение, то либо мы получим полигон пользователя, и сумму визитов и покупок и его уникальный user_id, 
или строки где каждому пользователю будет пренадлежать несколько строк с данными по визитам и покупкам и его уникальным адресам, если с одного адреса происходило несколько визитов и покупок, то в таком случаем можно сгруппировать данные (по принципу суммы).

In [204]:
import requests
import time
from bs4 import BeautifulSoup

### Задание 17
Будем парсить страницу со свежеми новостям на habr.com/ru/all/.

Вам необходимо собирать только те статьи, в которых встречается хотя бы одно требуемое ключевое слово. Эти слова определяем в начале кода в переменной, например:

KEYWORDS = ['python', 'парсинг']

Поиск вести по всей доступной preview-информации (это информация, доступная непосредственно с текущей страницы).

В итоге должен формироваться датафрейм вида: <дата> - <заголовок> - <ссылка>

### Решение

In [206]:
URL = 'https://habr.com/ru/all/'
words = ['python', 'парсинг']

In [207]:
res_main = requests.get(URL)
res_main.text

'<!DOCTYPE html>\n<html lang="ru" class="no-js">\n  <head>\n    <meta http-equiv="content-type" content="text/html; charset=utf-8" />\n<meta content=\'width=1024\' name=\'viewport\'>\n<title>Все публикации подряд / Хабр</title>\n\n\n\n  <meta property="fb:app_id" content="444736788986613" />\n<meta property="og:type" content="website"/>\n<meta property="fb:pages" content="472597926099084"/>\n<meta property="og:site_name" content="Хабр" />\n<link rel="image_src" href="https://habr.com/images/habr_ru.png" />\n<meta property="og:image" content="https://habr.com/images/habr_ru.png" />\n<meta property="og:image:width" content="1200" />\n<meta property="og:image:height" content="628" />\n<meta property="og:title" content="Все публикации подряд / Хабр"/>\n<meta property="og:description" content="Хабр — крупнейший в Европе ресурс для IT-специалистов. Сюда приходят обсудить новости индустрии и поделиться опытом."/>\n\n\n\n<meta name=\'yandex-verification\' content=\'71593b225aeafc4e\' />\n<meta

In [208]:
# работа с HTML страницы

soup_main = BeautifulSoup(res_main.text, 'html.parser')
soup_main

<!DOCTYPE html>

<html class="no-js" lang="ru">
<head>
<meta content="text/html; charset=utf-8" http-equiv="content-type"/>
<meta content="width=1024" name="viewport"/>
<title>Все публикации подряд / Хабр</title>
<meta content="444736788986613" property="fb:app_id">
<meta content="website" property="og:type"/>
<meta content="472597926099084" property="fb:pages"/>
<meta content="Хабр" property="og:site_name"/>
<link href="https://habr.com/images/habr_ru.png" rel="image_src"/>
<meta content="https://habr.com/images/habr_ru.png" property="og:image"/>
<meta content="1200" property="og:image:width"/>
<meta content="628" property="og:image:height"/>
<meta content="Все публикации подряд / Хабр" property="og:title"/>
<meta content="Хабр — крупнейший в Европе ресурс для IT-специалистов. Сюда приходят обсудить новости индустрии и поделиться опытом." property="og:description"/>
<meta content="71593b225aeafc4e" name="yandex-verification"/>
<meta content="unsafe-url" name="referrer"/>
<meta content

In [209]:
# добираемя до параметра который нам подскажет сколько страниц всего 

res_1 = soup_main.find('a', class_='toggle-menu__item-link toggle-menu__item-link_pagination toggle-menu__item-link_bordered').get('href')
res_1

'/ru/all/page50/'

In [210]:
# выводить число страниц со всеми страницами, то есть получаем что всего 50 страниц с постами

number_page = soup_main.find('a', class_='toggle-menu__item-link toggle-menu__item-link_pagination toggle-menu__item-link_bordered').get('href')
int(number_page[-3:-1])

50

In [211]:
# через метод range формируем цикл в рамках итераций которого будут формироваться url с постами

list(range(1,int(number_page[-3:-1])+1))

for x in range(1,int(number_page[-3:-1])+1):
    url_links = f'https://habr.com/ru/all/page{x}'
    print(url_links) 

https://habr.com/ru/all/page1
https://habr.com/ru/all/page2
https://habr.com/ru/all/page3
https://habr.com/ru/all/page4
https://habr.com/ru/all/page5
https://habr.com/ru/all/page6
https://habr.com/ru/all/page7
https://habr.com/ru/all/page8
https://habr.com/ru/all/page9
https://habr.com/ru/all/page10
https://habr.com/ru/all/page11
https://habr.com/ru/all/page12
https://habr.com/ru/all/page13
https://habr.com/ru/all/page14
https://habr.com/ru/all/page15
https://habr.com/ru/all/page16
https://habr.com/ru/all/page17
https://habr.com/ru/all/page18
https://habr.com/ru/all/page19
https://habr.com/ru/all/page20
https://habr.com/ru/all/page21
https://habr.com/ru/all/page22
https://habr.com/ru/all/page23
https://habr.com/ru/all/page24
https://habr.com/ru/all/page25
https://habr.com/ru/all/page26
https://habr.com/ru/all/page27
https://habr.com/ru/all/page28
https://habr.com/ru/all/page29
https://habr.com/ru/all/page30
https://habr.com/ru/all/page31
https://habr.com/ru/all/page32
https://habr.com/

In [212]:
# выводим текст с нужной страницы с постами

res_2 = requests.get(url_links)
res_2.text

'<!DOCTYPE html>\n<html lang="ru" class="no-js">\n  <head>\n    <meta http-equiv="content-type" content="text/html; charset=utf-8" />\n<meta content=\'width=1024\' name=\'viewport\'>\n<title>Все публикации подряд / Хабр</title>\n\n\n\n  <meta property="fb:app_id" content="444736788986613" />\n<meta property="og:type" content="website"/>\n<meta property="fb:pages" content="472597926099084"/>\n<meta property="og:site_name" content="Хабр" />\n<link rel="image_src" href="https://habr.com/images/habr_ru.png" />\n<meta property="og:image" content="https://habr.com/images/habr_ru.png" />\n<meta property="og:image:width" content="1200" />\n<meta property="og:image:height" content="628" />\n<meta property="og:title" content="Все публикации подряд / Хабр"/>\n<meta property="og:description" content="Хабр — крупнейший в Европе ресурс для IT-специалистов. Сюда приходят обсудить новости индустрии и поделиться опытом."/>\n\n\n\n<meta name=\'yandex-verification\' content=\'71593b225aeafc4e\' />\n<meta

In [213]:
# прасим нужную страницу с постами

soup_2 = BeautifulSoup(res_2.text, 'html.parser')
soup_2

<!DOCTYPE html>

<html class="no-js" lang="ru">
<head>
<meta content="text/html; charset=utf-8" http-equiv="content-type"/>
<meta content="width=1024" name="viewport"/>
<title>Все публикации подряд / Хабр</title>
<meta content="444736788986613" property="fb:app_id">
<meta content="website" property="og:type"/>
<meta content="472597926099084" property="fb:pages"/>
<meta content="Хабр" property="og:site_name"/>
<link href="https://habr.com/images/habr_ru.png" rel="image_src"/>
<meta content="https://habr.com/images/habr_ru.png" property="og:image"/>
<meta content="1200" property="og:image:width"/>
<meta content="628" property="og:image:height"/>
<meta content="Все публикации подряд / Хабр" property="og:title"/>
<meta content="Хабр — крупнейший в Европе ресурс для IT-специалистов. Сюда приходят обсудить новости индустрии и поделиться опытом." property="og:description"/>
<meta content="71593b225aeafc4e" name="yandex-verification"/>
<meta content="unsafe-url" name="referrer"/>
<meta content

In [214]:
# find  / findall функции bs4 ищит инфу на странице, article это тег, post - это класс, всегда возвращает список, в отличии от find

all_info = soup_2.find_all('article', class_='post')
all_info

[<article class="post post_preview" lang="ru">
 <header class="post__meta">
 <a class="post__user-info user-info" href="https://habr.com/ru/users/osidiustheemphatic/" title="Автор публикации">
 <img class="user-info__image-pic user-info__image-pic_small" height="24" src="//habrastorage.org/getpro/habr/avatars/006/1af/ab7/0061afab7d0d4b4979e365eee778f811.jpg" width="24"/>
 <span class="user-info__nickname user-info__nickname_small">osidiustheemphatic</span>
 </a>
 <span class="post__time">25 ноября 2020 в 21:56</span>
 </header>
 <h2 class="post__title">
 <a class="post__title_link" href="https://habr.com/ru/company/deutschetelekomitsolutions/blog/529934/">«Перебраться через забор» или история о том, как стать командой за три часа</a>
 </h2>
 <ul class="post__hubs inline-list">
 <li class="inline-list__item inline-list__item_hub">
 <a class="inline-list__item-link hub-link" href="https://habr.com/ru/company/deutschetelekomitsolutions/" onclick="if (typeof ga === 'function') { ga('send',

In [215]:
# добрались до ссылок

list_link = []
for article in all_info:
    my_link = article.find('a', class_= 'post__title_link').get('href')
    list_link.append(my_link)
    
list_link

['https://habr.com/ru/company/deutschetelekomitsolutions/blog/529934/',
 'https://habr.com/ru/company/pult/blog/529924/',
 'https://habr.com/ru/post/529926/',
 'https://habr.com/ru/post/529720/',
 'https://habr.com/ru/company/galssoftware/blog/529920/',
 'https://habr.com/ru/company/dataart/blog/529916/',
 'https://habr.com/ru/company/selectel/blog/529914/',
 'https://habr.com/ru/post/529908/',
 'https://habr.com/ru/post/529906/',
 'https://habr.com/ru/post/529904/',
 'https://habr.com/ru/post/529858/',
 'https://habr.com/ru/company/timeweb/blog/529894/',
 'https://habr.com/ru/post/529892/',
 'https://habr.com/ru/company/miro/blog/529890/',
 'https://habr.com/ru/company/itglobalcom/blog/529884/',
 'https://habr.com/ru/company/otus/blog/529886/',
 'https://habr.com/ru/company/JetBrains/blog/529634/',
 'https://habr.com/ru/post/529826/',
 'https://habr.com/ru/company/skillfactory/blog/528240/',
 'https://habr.com/ru/post/529874/']

In [216]:
# обернем наши действия в функцию , только одна страница

def all_links_1(URL):
    res = requests.get(URL)
    soup = BeautifulSoup(res.text, 'html.parser')
    all_info = soup.find_all('article', class_='post')
    list_link = []
    for article in all_info:
        my_link = article.find('a', class_= 'post__title_link').get('href')
        list_link.append(my_link)
    return list_link

In [217]:
# обернем наши действия в функцию , все страницы с постами

def all_links_2(URL):
    list_link = []
    res_1 = requests.get(URL)
    soup_1 = BeautifulSoup(res_1.text, 'html.parser')
    number_page = soup_1.find('a', class_='toggle-menu__item-link toggle-menu__item-link_pagination toggle-menu__item-link_bordered').get('href')
    for x in range(1,int(number_page[-3:-1])+1):
        url_links = f'https://habr.com/ru/all/page{x}'
        res_2 = requests.get(url_links)
        soup_2 = BeautifulSoup(res_2.text, 'html.parser')
        all_info = soup_2.find_all('article', class_='post')
        for article in all_info:
            my_link = article.find('a', class_= 'post__title_link').get('href')
            list_link.append(my_link)
    return list_link

In [218]:
all_links_1(URL)

['https://habr.com/ru/post/532778/',
 'https://habr.com/ru/post/532772/',
 'https://habr.com/ru/post/532770/',
 'https://habr.com/ru/post/532762/',
 'https://habr.com/ru/company/ruvds/blog/532604/',
 'https://habr.com/ru/company/gms/blog/532752/',
 'https://habr.com/ru/post/532746/',
 'https://habr.com/ru/post/532744/',
 'https://habr.com/ru/company/spbifmo/blog/532728/',
 'https://habr.com/ru/company/audiomania/blog/532026/',
 'https://habr.com/ru/post/532740/',
 'https://habr.com/ru/post/532732/',
 'https://habr.com/ru/post/532726/',
 'https://habr.com/ru/post/532716/',
 'https://habr.com/ru/post/532710/',
 'https://habr.com/ru/post/532622/',
 'https://habr.com/ru/post/532556/',
 'https://habr.com/ru/post/532698/',
 'https://habr.com/ru/post/532694/',
 'https://habr.com/ru/post/532692/']

In [219]:
# функция которая выводит список со всеми ссылками на посты
all_links_2(URL)

['https://habr.com/ru/post/532778/',
 'https://habr.com/ru/post/532772/',
 'https://habr.com/ru/post/532770/',
 'https://habr.com/ru/post/532762/',
 'https://habr.com/ru/company/ruvds/blog/532604/',
 'https://habr.com/ru/company/gms/blog/532752/',
 'https://habr.com/ru/post/532746/',
 'https://habr.com/ru/post/532744/',
 'https://habr.com/ru/company/spbifmo/blog/532728/',
 'https://habr.com/ru/company/audiomania/blog/532026/',
 'https://habr.com/ru/post/532740/',
 'https://habr.com/ru/post/532732/',
 'https://habr.com/ru/post/532726/',
 'https://habr.com/ru/post/532716/',
 'https://habr.com/ru/post/532710/',
 'https://habr.com/ru/post/532622/',
 'https://habr.com/ru/post/532556/',
 'https://habr.com/ru/post/532698/',
 'https://habr.com/ru/post/532694/',
 'https://habr.com/ru/post/532692/',
 'https://habr.com/ru/company/ruvds/blog/532330/',
 'https://habr.com/ru/company/skillfactory/blog/532688/',
 'https://habr.com/ru/company/selectel/blog/532680/',
 'https://habr.com/ru/company/audiom

In [220]:
# присваиваем переменной значение списка со всеми ссылками на посты
res_all_links = all_links_2(URL)
res_all_links

['https://habr.com/ru/post/532778/',
 'https://habr.com/ru/post/532772/',
 'https://habr.com/ru/post/532770/',
 'https://habr.com/ru/post/532762/',
 'https://habr.com/ru/company/ruvds/blog/532604/',
 'https://habr.com/ru/company/gms/blog/532752/',
 'https://habr.com/ru/post/532746/',
 'https://habr.com/ru/post/532744/',
 'https://habr.com/ru/company/spbifmo/blog/532728/',
 'https://habr.com/ru/company/audiomania/blog/532026/',
 'https://habr.com/ru/post/532740/',
 'https://habr.com/ru/post/532732/',
 'https://habr.com/ru/post/532726/',
 'https://habr.com/ru/post/532716/',
 'https://habr.com/ru/post/532710/',
 'https://habr.com/ru/post/532622/',
 'https://habr.com/ru/post/532556/',
 'https://habr.com/ru/post/532698/',
 'https://habr.com/ru/post/532694/',
 'https://habr.com/ru/post/532692/',
 'https://habr.com/ru/company/ruvds/blog/532330/',
 'https://habr.com/ru/company/skillfactory/blog/532688/',
 'https://habr.com/ru/company/selectel/blog/532680/',
 'https://habr.com/ru/company/audiom

In [222]:
# формируем итоговый дата-фрейм

our_table = pd.DataFrame()

for el in res_all_links:
    time.sleep(0.36)
    res_res = requests.get(el)
    res_soup = BeautifulSoup(res_res.text, 'html.parser')
    text_post = res_soup.find('article', class_='post') # в этой строке лежит HTML 
    for word in words:
        if word.lower() in text_post.text.lower():
            title_post = text_post.find('span', class_='post__title-text').text
            date_post = text_post.find('span', class_='post__time').text
            post_text = text_post.find('div', class_='post__text').text
            link_post = text_post.find('div', class_='post__text').get('data-io-article-url')
            row = {'date': date_post, 'title': title_post, 'link': link_post, 'text': post_text}
            our_table = pd.concat([our_table, pd.DataFrame([row])]) # в пустой датафрейм добавляем каждую новую строку с данными
            break
            

our_table 

date                                              title  \
0            вчера в 22:50                     Conways Game of life на Python   
0            вчера в 20:21  Russian AI Cup 2019. 4 место, почти не умея пр...   
0            вчера в 18:44  Визуализация данных по акциям дивидендных арис...   
0            вчера в 17:15          ALog: плюс один логгер для С++ приложений   
0            вчера в 15:57  Как я превратил свой стол с регулировкой высот...   
..                     ...                                                ...   
0   26 ноября 2020 в 07:30      Перевод вводной статьи от разработчиков D-BUS   
0   26 ноября 2020 в 03:32  Внедрение CI/CD: в чем основная задача пайплай...   
0   25 ноября 2020 в 20:44            Как перестать бояться и полюбить 1-wire   
0   25 ноября 2020 в 20:29              Андрей Терехов: от Фортрана до Питона   
0   25 ноября 2020 в 17:33  HyperBand и BOHB. Понимание современных алгори...   

    link                                               text  
0   None  Это мой первый пост, где я хочу рассказать про...  
0   None  Данная статья рассмотрит процесс моего участия...  
0   None   В этой статье я поделюсь некоторым опытом на ...  
0   None  Система логирования ALog первоначально разраба...  
0   None  В этой статье я покажу, как я превратил свой с...  
..   ...                                                ...  
0   None  Руководство по D-BUShttps://dbus.freedesktop.o...  
0   None  \nО своём опыте построения пайплайнов, правиль...  
0   None  Мне очень нравится протокол 1-wire своей прост...  
0   None  \n\r\nЭтой осенью Андрей Терехов — завкафедрой...  
0   None  Специально к старту курса «Машинное обучение» ...  

[147 rows x 4 columns]

In [223]:
# делал проверку, чтобы посмотреть что выводится в text_post

el = 'https://habr.com/ru/post/524082/'
res_res = requests.get(el)
res_soup = BeautifulSoup(res_res.text, 'html.parser')
text_post = res_soup.find('article', class_='post')


text_post

<article class="post post_full" id="post_524082" lang="ru">
<div class="post__wrapper">
<header class="post__meta">
<a class="post__user-info user-info" href="https://habr.com/ru/users/Nireko/" title="Автор публикации">
<img class="user-info__image-pic user-info__image-pic_small" height="24" src="//habrastorage.org/getpro/habr/avatars/861/7d1/ecb/8617d1ecb51007e47e11b1e0042cc9b8.png" width="24"/>
<span class="user-info__nickname user-info__nickname_small">Nireko</span>
</a>
<span class="post__time" data-time_published="2020-10-19T11:38Z">19 октября 2020 в 14:38</span>
</header>
<h1 class="post__title post__title_full">
<span class="post__title-text">Подключение Yandex Database к serverless телеграм боту на Yandex Functions</span>
</h1>
<ul class="post__hubs post__hubs_full-post inline-list">
<li class="inline-list__item inline-list__item_hub">
<a class="inline-list__item-link hub-link" href="https://habr.com/ru/hub/python/" onclick="if (typeof ga === 'function') { ga('send', 'event', '

### Задание 18

Обязательная часть
Написать скрипт, который будет проверять список e-mail адресов на утечку при помощи сервиса Avast Hack Ckeck. Список email-ов задаем переменной в начале кода:
EMAIL = [xxx@x.ru, yyy@y.com]

В итоге должен формироваться датафрейм со столбцами: <почта> - <дата утечки> - <источник утечки> - <описание утечки>

Подсказка: сервис работает при помощи "скрытого" API. Внимательно изучите post-запросы.

### Решение

In [228]:
import json

In [229]:
EMAIL = ['xxx@x.ru', 'yyy@y.com']

In [230]:
URL = 'https://identityprotection.avast.com/v1/web/query/site-breaches/unauthorized-data'

In [231]:
params = {
    'emailAddresses': EMAIL
}

In [232]:
headers = {
    'Vaar-Version': '0',
    'Vaar-Header-App-Product': 'hackcheck-web-avast'
}

In [233]:
res = requests.post(URL, json=params, headers=headers)
res

<Response [200]>

In [234]:
# вариант в одну ячейку

EMAIL = ['xxx@x.ru', 'yyy@y.com']
URL = 'https://identityprotection.avast.com/v1/web/query/site-breaches/unauthorized-data'
params = {
    'emailAddresses': EMAIL
}
headers = {
    'Vaar-Version': '0',
    'Vaar-Header-App-Product': 'hackcheck-web-avast'
}
res= requests.post(URL, json=params, headers=headers)
res

<Response [200]>

In [235]:
res.text

'{"breaches":{"17110":{"breachId":17110,"site":"azcentral.com","recordsCount":705538,"description":"At an unconfirmed date, online Arizona newspaper \\"AZ Central\\" was allegedly breached. The stolen data contains passwords and email addresses. This breach is being publicly shared on the internet.","publishDate":"2020-01-03T00:00:00Z","statistics":{"usernames":0,"passwords":702971,"emails":705538}},"17670":{"breachId":17670,"site":"wishbone.io","recordsCount":37475476,"description":"In January 2020, the online poll website Wishbone was allegedly breached. The stolen data contains usernames, passwords, email addresses and additional personal information. This breach is being privately shared on the internet.","publishDate":"2020-05-28T00:00:00Z","statistics":{"usernames":31111127,"passwords":10648734,"emails":10234249}},"13094":{"breachId":13094,"site":"myheritage.com","recordsCount":110041653,"description":"In October 2017, a customer database belonging to online genealogy platform My

In [236]:
# выяснял глубину вложенности словарей

for key, value in res.json()['summary'].items():
    #print(key)
    #print(value)
    email = key
    for key_2, value_2 in value.items():
        #print(key_2)
        #print(value_2)
        for el in value_2:
            print(el)

3176
12
3
3164
2961
15
17110
17670
13094
16768
41
2
3587
17009
3
13662
3520
15
3669


In [237]:
# выяснял глубину вложенности словарей

for key_4, value_4 in res.json()['breaches'].items():
    print(key_4)
    #print(value_4)
    site = value_4['site']
    description = value_4['description']
    date = value_4['publishDate']

print(site)
print(description)
print(date)

17110
17670
13094
16768
3176
41
2
3587
17009
12
3
13662
3520
3164
2961
15
3669
youku.com
Youku is a large Chinese video content company whose user database was allegedly breached in 2016. The compromised data later appeared for sale on a dark web marketplace April 2017. The dark web listing advertised the sale of 100M user accounts.
2017-03-24T00:00:00Z


In [238]:
my_df = pd.DataFrame()

for key, value in res.json()['summary'].items():
    for key_2, value_2 in value.items():
        for el in value_2:
            el
        for key_4, value_4 in res.json()['breaches'].items():
            site = value_4['site']
            description = value_4['description']
            date = value_4['publishDate']
            email = key
            res_row = {'email': email, 'date': date, 'source': site, 'description': description}
            my_df = pd.concat([my_df, pd.DataFrame([res_row])])

my_df

email                  date            source  \
0   xxx@x.ru  2020-01-03T00:00:00Z     azcentral.com   
0   xxx@x.ru  2020-05-28T00:00:00Z       wishbone.io   
0   xxx@x.ru  2017-11-04T00:00:00Z    myheritage.com   
0   xxx@x.ru  2019-06-13T00:00:00Z         canva.com   
0   xxx@x.ru  2017-02-14T00:00:00Z    parapa.mail.ru   
0   xxx@x.ru  2016-10-24T00:00:00Z       dropbox.com   
0   xxx@x.ru  2016-10-21T00:00:00Z      linkedin.com   
0   xxx@x.ru  2017-03-01T00:00:00Z      rayli.com.cn   
0   xxx@x.ru  2019-10-17T00:00:00Z         zynga.com   
0   xxx@x.ru  2016-10-29T00:00:00Z            vk.com   
0   xxx@x.ru  2016-10-21T00:00:00Z         adobe.com   
0   xxx@x.ru  2018-02-18T00:00:00Z        netlog.com   
0   xxx@x.ru  2017-03-15T00:00:00Z    globalreach.eu   
0   xxx@x.ru  2017-02-14T00:00:00Z     cfire.mail.ru   
0   xxx@x.ru  2017-01-31T00:00:00Z  cdprojektred.com   
0   xxx@x.ru  2016-10-23T00:00:00Z         imesh.com   
0   xxx@x.ru  2017-03-24T00:00:00Z         youku.com   
0  yyy@y.com  2020-01-03T00:00:00Z     azcentral.com   
0  yyy@y.com  2020-05-28T00:00:00Z       wishbone.io   
0  yyy@y.com  2017-11-04T00:00:00Z    myheritage.com   
0  yyy@y.com  2019-06-13T00:00:00Z         canva.com   
0  yyy@y.com  2017-02-14T00:00:00Z    parapa.mail.ru   
0  yyy@y.com  2016-10-24T00:00:00Z       dropbox.com   
0  yyy@y.com  2016-10-21T00:00:00Z      linkedin.com   
0  yyy@y.com  2017-03-01T00:00:00Z      rayli.com.cn   
0  yyy@y.com  2019-10-17T00:00:00Z         zynga.com   
0  yyy@y.com  2016-10-29T00:00:00Z            vk.com   
0  yyy@y.com  2016-10-21T00:00:00Z         adobe.com   
0  yyy@y.com  2018-02-18T00:00:00Z        netlog.com   
0  yyy@y.com  2017-03-15T00:00:00Z    globalreach.eu   
0  yyy@y.com  2017-02-14T00:00:00Z     cfire.mail.ru   
0  yyy@y.com  2017-01-31T00:00:00Z  cdprojektred.com   
0  yyy@y.com  2016-10-23T00:00:00Z         imesh.com   
0  yyy@y.com  2017-03-24T00:00:00Z         youku.com   

                                         description  
0  At an unconfirmed date, online Arizona newspap...  
0  In January 2020, the online poll website Wishb...  
0  In October 2017, a customer database belonging...  
0  In May 2019, graphic-design site Canva's datab...  
0  In July and August 2016, two criminals execute...  
0  Cloud storage company Dropbox suffered a major...  
0  In 2012, online professional networking platfo...  
0  On an unconfirmed date, Chinese gossip site Ra...  
0  In September 2019, the game developer Zynga wa...  
0  Popular Russian social networking platform VKo...  
0  In October of 2013, criminals penetrated Adobe...  
0  Netlog (formerly known as Facebox and Bingbox)...  
0  In 2016, Global Reach Technology's database wa...  
0  In July and August of 2016, two criminals carr...  
0  In March 2016, CDProjektRed.com.com's forum da...  
0  In June 2016, a cache of over 51 million user ...  
0  Youku is a large Chinese video content company...  
0  At an unconfirmed date, online Arizona newspap...  
0  In January 2020, the online poll website Wishb...  
0  In October 2017, a customer database belonging...  
0  In May 2019, graphic-design site Canva's datab...  
0  In July and August 2016, two criminals execute...  
0  Cloud storage company Dropbox suffered a major...  
0  In 2012, online professional networking platfo...  
0  On an unconfirmed date, Chinese gossip site Ra...  
0  In September 2019, the game developer Zynga wa...  
0  Popular Russian social networking platform VKo...  
0  In October of 2013, criminals penetrated Adobe...  
0  Netlog (formerly known as Facebox and Bingbox)...  
0  In 2016, Global Reach Technology's database wa...  
0  In July and August of 2016, two criminals carr...  
0  In March 2016, CDProjektRed.com.com's forum da...  
0  In June 2016, a cache of over 51 million user ...  
0  Youku is a large Chinese video content company...

In [239]:
# оптимальный вариант

EMAIL = ['xxx@x.ru', 'yyy@y.com']
url = 'https://identityprotection.avast.com/v1/web/query/site-breaches/unauthorized-data'
headers = {'Vaar-Header-App-Product': 'hackcheck-web-avast','Vaar-Version': '0'}
adresses = {'emailAddresses': EMAIL}
res = requests.post(url, json=adresses, headers=headers)
df = pd.DataFrame(res.json()['breaches']).T.reset_index() 
df = df[['site', 'description', 'publishDate']]

In [240]:
df.head()

site                                        description  \
0   azcentral.com  At an unconfirmed date, online Arizona newspap...   
1     wishbone.io  In January 2020, the online poll website Wishb...   
2  myheritage.com  In October 2017, a customer database belonging...   
3       canva.com  In May 2019, graphic-design site Canva's datab...   
4  parapa.mail.ru  In July and August 2016, two criminals execute...   

            publishDate  
0  2020-01-03T00:00:00Z  
1  2020-05-28T00:00:00Z  
2  2017-11-04T00:00:00Z  
3  2019-06-13T00:00:00Z  
4  2017-02-14T00:00:00Z